scapy/layers/http.py

```diff
@@ -0,0 +1,471 @@
+                   StrField("Headers", None, fmt="H")]
+
+    def do_dissect(self, s):
+        """From the HTTP packet string, populate the scapy object"""
+        first_line, body = _dissect_headers(self, s)
+        Method, Path, HTTPVersion = re.split(br"\s+", first_line)
+        self.setfieldval('Method', Method)
+        self.setfieldval('Path', Path)
+        self.setfieldval('Http_Version', HTTPVersion)
+        return body
+
+
+class HTTPResponse(_HTTPContent):
+    name = "HTTP Response"
+    fields_desc = [StrField("Status_Line", None, fmt="H"),
+                   StrField("Accept_Ranges", None, fmt="H"),
+                   StrField("Age", None, fmt="H"),
+                   StrField("E_Tag", None, fmt="H"),
+                   StrField("Location", None, fmt="H"),
+                   StrField("Proxy_Authenticate", None, fmt="H"),
+                   StrField("Retry_After", None, fmt="H"),
+                   StrField("Server", None, fmt="H"),
+                   StrField("Vary", None, fmt="H"),
+                   StrField("WWW_Authenticate", None, fmt="H"),
+                   StrField("Cache_Control", None, fmt="H"),
+                   StrField("Connection", None, fmt="H"),
+                   StrField("Date", None, fmt="H"),
+                   StrField("Pragma", None, fmt="H"),
+                   StrField("Trailer", None, fmt="H"),
+                   StrField("Transfer_Encoding", None, fmt="H"),
+                   StrField("Upgrade", None, fmt="H"),
+                   StrField("Via", None, fmt="H"),
+                   StrField("Warning", None, fmt="H"),
+                   StrField("Keep_Alive", None, fmt="H"),
+                   StrField("Allow", None, fmt="H"),
+                   StrField("Content_Encoding", None, fmt="H"),
+                   StrField("Content_Language", None, fmt="H"),
+                   StrField("Content_Length", None, fmt="H"),
+                   StrField("Content_Location", None, fmt="H"),
+                   StrField("Content_MD5", None, fmt="H"),
+                   StrField("Content_Range", None, fmt="H"),
+                   StrField("Content_Type", None, fmt="H"),
+                   StrField("Expires", None, fmt="H"),
+                   StrField("Last_Modified", None, fmt="H"),
+                   StrField("Headers", None, fmt="H")]
+
+    def do_dissect(self, s):
+        ''' From the HTTP packet string, populate the scapy object '''
+        first_line, body = _dissect_headers(self, s)
+        self.setfieldval('Status_Line', first_line)
+        return body
+
+
+class HTTP(Packet):
+    name = "HTTP"
+    fields_desc = []
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
+        if _pkt and False:
+            # XXX TODO
+            from scapy.contrib.contrib.http2 import H2Frame
+            return H2Frame
+        return cls
+
+    # tcp_reassemble is used by TCPSession in session.py
+    @classmethod
+    def tcp_reassemble(cls, frags, metadata):
```

Comment:
Is it possible to split this function using other functions? This will make it easier to read and review.

Reply:
It might be a bit hard. This function supports several decompression method, and though there are different, separating it would probably mean duplicating some code.

I've added quite a bunch of comments, but I could add more if it helps

Reply:
I've extended the comments to be as explicit as possible. Hope it helps 😕

Reply:
It sure helps. Don't you think that the steps can be moved to separate functions?

Reply:
I don't think that splitting the code will make it easier :/ It doesn't really add anything, and we'll have to scroll to find each part.

Also I'll have to add some extra logic for the returns..

I'd rather keep it as one if you don't mind :-)

Reply:
I've ended up reworking the function entirely